File: searching/python/searching_runner.py
"""
Benchmark de algoritmos de busqueda — SOLO PYTHON (Punto 2).

- Lee datos desde archivos pre-generados en data_searching/ (no regenera).
- Mide UNICAMENTE el tiempo de busqueda (sin I/O).
- Exporta resultados en results_searching/ (CSV/JSON) y grafica (PNG).

Uso:
    python searching/python/searching_runner.py
"""
from __future__ import annotations  # Permite usar anotaciones de tipo modernas en versiones anteriores de Python

import csv        # Para escribir archivos CSV con los resultados
import json       # Para escribir archivos JSON con los resultados
import subprocess # Para ejecutar subprocesos externos (el script generador de datos)
import sys        # Para manipular el path del sistema y el intérprete actual
import time       # Para medir tiempos de ejecución con alta precisión
from pathlib import Path  # Para manejar rutas de archivos de forma multiplataforma

_SEARCH_DIR = Path(__file__).resolve().parent  # Directorio donde vive este script
if str(_SEARCH_DIR) not in sys.path:           # Si ese directorio no está en el path de Python...
    sys.path.insert(0, str(_SEARCH_DIR))       # ...lo agrega al inicio para poder importar módulos locales

# Importa cada algoritmo de búsqueda desde sus módulos locales
from binary_search import binarySearch    # Importa Búsqueda Binaria
from ternary_search import ternarySearch  # Importa Búsqueda Ternaria
from jump_search import jumpSearch        # Importa Búsqueda por Saltos

ROOT = Path(__file__).resolve().parents[2]              # Sube 2 niveles desde el script para llegar a la raíz del proyecto
DATA_DIR = ROOT / "data_searching"                      # Carpeta donde se encuentran los archivos de datos de entrada
RESULTS_DIR = ROOT / "results_searching"                # Carpeta donde se guardarán los resultados del benchmark
GEN_SCRIPT = ROOT / "scripts" / "generate_searching_data.py"  # Ruta al script que genera los datos de prueba

# Lista de tamaños a probar: cada tupla es (cantidad de elementos, archivo de arreglo, archivo de queries)
SIZES: list[tuple[int, str, str]] = [
    (10_000,    "searching_10000.txt",   "queries_10000.txt"),   # Arreglo pequeño de 10.000 elementos
    (100_000,   "searching_100000.txt",  "queries_100000.txt"),  # Arreglo mediano de 100.000 elementos
    (1_000_000, "searching_1000000.txt", "queries_1000000.txt"), # Arreglo grande de 1.000.000 elementos
]

# Lista de algoritmos a evaluar: cada tupla es (nombre, complejidad, función de búsqueda)
ALGORITHMS: list[tuple[str, str, object]] = [
    ("Binary Search",  "O(log n)",   binarySearch),                    # Búsqueda binaria clásica
    ("Ternary Search", "O(log n)",   ternarySearch),                   # Búsqueda ternaria (divide en 3 partes)
    ("Jump Search",    "O(sqrt(n))", lambda a, x: jumpSearch(a, x)),   # Búsqueda por saltos envuelta en lambda para uniformar la firma
]


def ensure_data() -> None:
    subprocess.run([sys.executable, str(GEN_SCRIPT)], check=True, cwd=str(ROOT))
    # Ejecuta el script generador usando el mismo intérprete Python actual, desde la raíz del proyecto


def load_ints(path: Path) -> list[int]:
    with path.open("r", encoding="ascii") as f:          # Abre el archivo en modo lectura con codificación ASCII
        return [int(line) for line in f if line.strip()] # Lee cada línea no vacía y la convierte a entero, retornando la lista


def run_python_benchmarks() -> list[dict]:
    rows: list[dict] = []  # Lista que acumulará los resultados de cada combinación algoritmo/tamaño

    for n, arr_file, q_file in SIZES:           # Itera sobre cada tamaño y sus archivos correspondientes
        arr_path = DATA_DIR / arr_file           # Construye la ruta completa al archivo del arreglo ordenado
        q_path   = DATA_DIR / q_file            # Construye la ruta completa al archivo de queries

        print(f"\n  -- n = {n:,} --")           # Imprime un separador visual con el tamaño actual formateado con comas
        print(f"    Cargando arreglo: {arr_file}") # Indica qué archivo de arreglo se está cargando
        arr = load_ints(arr_path)               # Carga los enteros del arreglo ordenado desde el archivo
        print(f"    Cargando queries: {q_file}") # Indica qué archivo de queries se está cargando
        queries = load_ints(q_path)             # Carga los valores a buscar desde el archivo de queries

        for name, complexity, fn in ALGORITHMS:  # Itera sobre cada algoritmo de búsqueda definido
            print(f"    [{name}] ejecutando ...", flush=True)  # Indica que el algoritmo está comenzando, vaciando el buffer

            # Medir SOLO busqueda (loop de queries)
            t0 = time.perf_counter()             # Registra el tiempo exacto de inicio de todas las búsquedas
            checksum = 0                         # Acumulador de resultados para verificar corrección y evitar optimizaciones del intérprete
            for q in queries:                    # Itera sobre cada valor a buscar en el arreglo
                checksum += fn(arr, q)           # Ejecuta la búsqueda y acumula el índice retornado en el checksum
            secs = time.perf_counter() - t0      # Calcula el tiempo total de todas las búsquedas en segundos

            label = f"{secs*1000:.1f} ms" if secs < 1 else f"{secs:.3f} s"  # Formatea el tiempo: ms si < 1s, segundos si ≥ 1s
            print(f"    [{name}] OK {label} (checksum={checksum})") # Imprime el resultado exitoso con tiempo y checksum

            rows.append({                        # Agrega un diccionario con los resultados de esta ejecución
                "algorithm":   name,             # Nombre del algoritmo ejecutado
                "language":    "Python",         # Lenguaje de programación usado
                "n":           n,                # Tamaño del arreglo sobre el que se buscó
                "timeSeconds": secs,             # Tiempo total de todas las búsquedas en segundos
                "complexity":  complexity,       # Complejidad teórica del algoritmo
            })

    return rows  # Retorna todos los resultados acumulados


def write_csv(path: Path, rows: list[dict]) -> None:
    fieldnames = ["algorithm", "language", "n", "timeSeconds", "complexity"]  # Define el orden de las columnas del CSV
    with path.open("w", newline="", encoding="utf-8") as f:  # Abre el archivo CSV para escritura en UTF-8
        w = csv.DictWriter(f, fieldnames=fieldnames)          # Crea un escritor de CSV basado en diccionarios
        w.writeheader()                                        # Escribe la fila de encabezados con los nombres de columna
        for row in rows:                                       # Itera sobre cada resultado del benchmark
            w.writerow({                                       # Escribe una fila con los datos formateados
                "algorithm":   row["algorithm"],               # Nombre del algoritmo
                "language":    row["language"],                # Lenguaje usado
                "n":           row["n"],                       # Tamaño del arreglo
                "timeSeconds": f"{row['timeSeconds']:.9f}",   # Tiempo con 9 decimales de precisión
                "complexity":  row["complexity"],              # Complejidad teórica
            })


def write_json(path: Path, rows: list[dict]) -> None:
    with path.open("w", encoding="utf-8") as f:              # Abre el archivo JSON para escritura en UTF-8
        json.dump(rows, f, indent=2, ensure_ascii=False)     # Serializa la lista de resultados con indentación y soporte UTF-8


def plot_bars(rows: list[dict], out_png: Path) -> None:
    try:
        import matplotlib.pyplot as plt  # Importa matplotlib para graficar (dentro del try por si no está instalado)
        import numpy as np               # Importa numpy para manejar posiciones numéricas de las barras
    except ImportError as e:
        print(f"matplotlib no instalado - se omite el grafico. ({e})")  # Si no están instalados, avisa y sale
        return                                                            # Sale de la función sin graficar

    order = [a[0] for a in ALGORITHMS]  # Extrae solo los nombres de los algoritmos en el orden definido
    fig, axes = plt.subplots(1, 3, figsize=(20, 6), sharey=False)  # Crea una figura con 3 subgráficos lado a lado (uno por tamaño)
    fig.suptitle(                        # Agrega un título general a toda la figura
        "Benchmark Algoritmos de Busqueda - Python\n"
        "Tiempo de busqueda puro (sin I/O)",
        fontsize=13,
        fontweight="bold",
    )

    for ax, (n, _, _) in zip(axes, SIZES):  # Itera sobre cada subgráfico y su tamaño (ignora los nombres de archivo con _)
        mp = {r["algorithm"]: r for r in rows if r["n"] == n and r["language"] == "Python"}
        # Filtra los resultados Python para este tamaño n y los indexa por nombre de algoritmo
        x = np.arange(len(order))            # Crea posiciones numéricas en el eje X para cada algoritmo
        vals = [mp[k]["timeSeconds"] if k in mp else 0 for k in order]  # Obtiene el tiempo de cada algoritmo (0 si no existe)
        bars = ax.bar(x, vals, 0.55, label="Python", color="#4C72B0", edgecolor="white", linewidth=0.6)
        # Dibuja las barras azules con ancho 0.55 y borde blanco fino
        ax.set_title(f"n = {n:,}", fontsize=11, fontweight="bold")      # Establece el título del subgráfico con el tamaño n
        ax.set_xticks(x)                                                 # Ubica las marcas del eje X en las posiciones de los algoritmos
        ax.set_xticklabels(order, rotation=20, ha="right", fontsize=9)  # Etiqueta el eje X con los nombres rotados 20°
        ax.set_ylabel("Tiempo (s)")                                      # Etiqueta el eje Y como "Tiempo (s)"
        ax.legend()                                                       # Muestra la leyenda del subgráfico

        for bar, v in zip(bars, vals):  # Itera sobre cada barra y su valor para añadir anotaciones de tiempo
            if v <= 0:                  # Si el valor es 0 o negativo (no ejecutado o error)...
                continue                # ...no agrega etiqueta sobre esa barra
            h  = bar.get_height()      # Obtiene la altura de la barra (igual al valor de tiempo)
            cx = bar.get_x() + bar.get_width() / 2  # Calcula el centro horizontal de la barra
            label = f"{v*1000:.2f} ms" if v < 1 else f"{v:.3f} s"  # Formatea el tiempo: ms si < 1s, segundos si ≥ 1s
            ax.annotate(label, xy=(cx, h), xytext=(0, 3), textcoords="offset points",
                        ha="center", va="bottom", fontsize=8)
            # Agrega la etiqueta de tiempo 3 puntos arriba del tope de la barra, centrada horizontalmente

    plt.tight_layout()                                   # Ajusta automáticamente los márgenes para evitar solapamientos
    plt.savefig(out_png, dpi=160, bbox_inches="tight")  # Guarda el gráfico como PNG con 160 DPI y sin recortes
    plt.close()                                          # Cierra la figura para liberar memoria
    print(f"\n  Grafico guardado en {out_png}")          # Informa en consola dónde se guardó el gráfico


def main() -> None:
    RESULTS_DIR.mkdir(parents=True, exist_ok=True)  # Crea la carpeta de resultados si no existe (y sus padres también)
    print("=" * 60)                                 # Imprime una línea decorativa de separación
    print("  BENCHMARK DE BUSQUEDA - PYTHON")       # Imprime el título del benchmark
    print("=" * 60)                                 # Imprime otra línea decorativa

    print("\n[1] Verificando/generando archivos de datos ...") # Informa el paso 1
    ensure_data()                                              # Genera los archivos de datos si no existen

    print("\n[2] Ejecutando benchmarks Python ...")  # Informa el paso 2
    rows = run_python_benchmarks()                  # Ejecuta todos los algoritmos y recopila resultados
    rows.sort(key=lambda r: (r["n"], r["algorithm"]))  # Ordena los resultados por tamaño n y luego por nombre de algoritmo

    csv_path  = RESULTS_DIR / "searching_benchmark_python.csv"  # Ruta de salida para el archivo CSV
    json_path = RESULTS_DIR / "searching_benchmark_python.json" # Ruta de salida para el archivo JSON
    png_path  = RESULTS_DIR / "searching_benchmark_python.png"  # Ruta de salida para el gráfico PNG

    write_csv(csv_path, rows)    # Exporta los resultados al archivo CSV
    write_json(json_path, rows)  # Exporta los resultados al archivo JSON
    print("\n[3] Resultados exportados:")            # Informa el paso 3
    print(f"    CSV  -> {csv_path}")                # Muestra la ruta del CSV generado
    print(f"    JSON -> {json_path}")               # Muestra la ruta del JSON generado

    plot_bars(rows, png_path)  # Genera y guarda el gráfico de barras comparativo

    print("\n" + "=" * 60)                          # Imprime una línea decorativa de cierre
    print("  BENCHMARK PYTHON COMPLETADO")          # Mensaje final de éxito
    print("=" * 60)                                 # Última línea decorativa


if __name__ == "__main__":  # Solo ejecuta main() si este script es el punto de entrada (no si es importado)
    main()                  # Llama a la función principal para arrancar el benchmark

File: searching/java/SearchingRunner.java
import java.io.BufferedReader;           // Para leer archivos de texto línea por línea de forma eficiente
import java.io.IOException;              // Para manejar excepciones de entrada/salida
import java.io.PrintStream;              // Para imprimir en consola con control de buffer y codificación
import java.nio.charset.StandardCharsets; // Para especificar codificaciones de texto estándar (UTF-8)
import java.nio.file.Files;              // Para operaciones modernas sobre archivos (leer, escribir, verificar existencia)
import java.nio.file.Path;               // Para representar rutas de archivos de forma multiplataforma
import java.util.ArrayList;              // Para usar listas dinámicas que acumulan los resultados
import java.util.Arrays;                 // Para operaciones sobre arreglos (copiar, redimensionar)
import java.util.List;                   // Interfaz base para listas en Java
import java.util.Locale;                 // Para formatear números con punto decimal en lugar de coma

/**
 * Benchmark de algoritmos de busqueda - SOLO JAVA (Punto 2).
 *
 * - Lee datos desde data_searching/ (arreglos ordenados + queries).
 * - Mide UNICAMENTE el tiempo de busqueda (sin I/O).
 * - Exporta JSON + CSV.
 *
 * Uso:
 *   java -cp searching/java SearchingRunner <dir_datos> <salida.json>
 *
 * Ejemplo:
 *   javac searching/java/*.java
 *   java -cp searching/java SearchingRunner data_searching results_searching/searching_benchmark_java.json
 */
public class SearchingRunner {

    // Tamaños de arreglos a evaluar en el benchmark
    private static final int[] SIZES = { 10_000, 100_000, 1_000_000 };
    // Nombres de los archivos de arreglos ordenados correspondientes a cada tamaño
    private static final String[] ARR_FILES = {
        "searching_10000.txt",
        "searching_100000.txt",
        "searching_1000000.txt"
    };
    // Nombres de los archivos de queries (valores a buscar) correspondientes a cada tamaño
    private static final String[] Q_FILES = {
        "queries_10000.txt",
        "queries_100000.txt",
        "queries_1000000.txt"
    };

    @FunctionalInterface                   // Marca que esta interfaz tiene exactamente un método abstracto (permite lambdas)
    private interface SearchFn {
        int search(int[] a, int x);        // Contrato que deben cumplir todos los algoritmos: recibe el arreglo y el valor a buscar, retorna el índice encontrado
    }

    private record AlgoCase(String name, String complexity, SearchFn fn) {}
    // Record inmutable que agrupa el nombre del algoritmo, su complejidad teórica y la función que lo ejecuta

    private record ResultRow(String algorithm, String language, int n, double timeSeconds, String complexity) {}
    // Record inmutable que representa una fila de resultado con todos los datos de una ejecución

    public static void main(String[] args) throws IOException {
        if (args.length < 2) {                                                         // Verifica que se pasaron al menos 2 argumentos por línea de comandos
            System.err.println("Uso: java SearchingRunner <dir_datos> <salida.json>"); // Muestra el uso correcto si faltan argumentos
            System.exit(1);                                                            // Termina el programa con código de error 1
        }

        Path dataDir = Path.of(args[0]);                                // Convierte el primer argumento en la ruta del directorio de datos
        Path outPath = Path.of(args[1]);                                // Convierte el segundo argumento en la ruta del archivo JSON de salida
        if (outPath.getParent() != null) outPath.getParent().toFile().mkdirs();
        // Si el archivo de salida tiene directorio padre, lo crea si no existe (evita error al escribir)

        PrintStream out = new PrintStream(System.out, true, "UTF-8");  // Crea un PrintStream con autoflush activado y codificación UTF-8

        out.println("============================================================"); // Línea decorativa de separación
        out.println("  BENCHMARK DE BUSQUEDA - JAVA");                               // Título del benchmark
        out.println("============================================================"); // Línea decorativa de separación

        AlgoCase[] cases = {
            new AlgoCase("Binary Search",  "O(log n)",   BinarySearch::search),
            // Búsqueda binaria clásica usando referencia al método estático search
            new AlgoCase("Ternary Search", "O(log n)",   TernarySearch::search),
            // Búsqueda ternaria (divide en 3 partes) usando referencia al método estático search
            new AlgoCase("Jump Search",    "O(sqrt(n))", JumpSearch::search),
            // Búsqueda por saltos de raíz cuadrada usando referencia al método estático search
        };

        List<ResultRow> rows = new ArrayList<>();       // Lista que acumulará todas las filas de resultados para el CSV
        StringBuilder json = new StringBuilder("[\n"); // Inicia la construcción del JSON como array abierto
        boolean first = true;                           // Bandera para controlar si es el primer elemento (para omitir la coma inicial)

        for (int si = 0; si < SIZES.length; si++) {    // Itera sobre cada tamaño de arreglo definido
            int n = SIZES[si];                          // Obtiene el tamaño actual
            Path arrPath = dataDir.resolve(ARR_FILES[si]); // Construye la ruta completa al archivo del arreglo ordenado
            Path qPath   = dataDir.resolve(Q_FILES[si]);   // Construye la ruta completa al archivo de queries

            if (!Files.exists(arrPath) || !Files.exists(qPath)) { // Verifica que ambos archivos existan antes de continuar
                System.err.println("  [ERROR] Faltan archivos de datos en: " + dataDir);       // Informa que faltan archivos
                System.err.println("  Necesarios: " + arrPath + " y " + qPath);                // Muestra exactamente qué archivos faltan
                System.err.println("  Ejecuta primero: python scripts/generate_searching_data.py"); // Indica cómo generarlos
                System.exit(1);                                                                 // Termina el programa con error
            }

            out.println("\n  -- n = " + String.format("%,d", n) + " --"); // Imprime el separador visual con el tamaño formateado con comas
            out.print("    Cargando arreglo ...");           // Indica que está comenzando la carga del arreglo ordenado
            out.flush();                                     // Vacía el buffer para que el mensaje aparezca de inmediato
            int[] arr = loadInts(arrPath, n);                // Carga hasta n enteros desde el archivo del arreglo ordenado
            out.println(" OK (" + arr.length + " elementos)"); // Confirma la carga exitosa con la cantidad real de elementos

            out.print("    Cargando queries ...");           // Indica que está comenzando la carga de los valores a buscar
            out.flush();                                     // Vacía el buffer para que el mensaje aparezca de inmediato
            int[] queries = loadInts(qPath, Integer.MAX_VALUE); // Carga TODOS los queries del archivo (sin límite de cantidad)
            out.println(" OK (" + queries.length + " consultas)"); // Confirma la carga exitosa con la cantidad de queries

            for (AlgoCase c : cases) {                       // Itera sobre cada algoritmo de búsqueda definido
                out.println("    [" + c.name() + "] ejecutando ..."); // Indica que el algoritmo está por comenzar
                out.flush();                                 // Vacía el buffer para que el mensaje aparezca de inmediato

                long checksum = 0;                           // Acumulador de resultados para verificar corrección y evitar optimizaciones del compilador
                long t0 = System.nanoTime();                 // Registra el tiempo exacto de inicio de todas las búsquedas
                for (int q : queries) {                      // Itera sobre cada valor a buscar en el arreglo
                    checksum += c.fn().search(arr, q);       // Ejecuta la búsqueda y acumula el índice retornado en el checksum
                }
                double secs = (System.nanoTime() - t0) / 1_000_000_000.0; // Calcula el tiempo total de todas las búsquedas en segundos

                String label = secs < 1.0
                    ? String.format(Locale.US, "%.2f ms", secs * 1000) // Si tardó menos de 1s, muestra en milisegundos con 2 decimales
                    : String.format(Locale.US, "%.3f s",  secs);       // Si tardó 1s o más, muestra en segundos con 3 decimales
                out.println("    [" + c.name() + "] OK " + label + " (checksum=" + checksum + ")"); // Imprime resultado con tiempo y checksum
                out.flush();                                 // Vacía el buffer para que el resultado aparezca de inmediato

                rows.add(new ResultRow(c.name(), "Java", n, secs, c.complexity())); // Agrega la fila de resultado a la lista para el CSV

                if (!first) json.append(",\n"); // Si no es el primer elemento, agrega coma separadora antes del nuevo objeto JSON
                first = false;                  // Marca que ya no es el primer elemento
                json.append("  {");             // Abre el objeto JSON para este resultado
                json.append("\"algorithm\":\"").append(esc(c.name())).append("\",");              // Agrega el nombre del algoritmo escapado
                json.append("\"language\":\"Java\",");                                             // Agrega el lenguaje fijo como "Java"
                json.append("\"n\":").append(n).append(",");                                       // Agrega el tamaño del arreglo
                json.append("\"timeSeconds\":").append(String.format(Locale.US, "%.9f", secs)).append(","); // Agrega el tiempo con 9 decimales
                json.append("\"complexity\":\"").append(esc(c.complexity())).append("\"");         // Agrega la complejidad teórica escapada
                json.append("}");               // Cierra el objeto JSON
            }
        }

        json.append("\n]\n");                                            // Cierra el array JSON y agrega salto de línea final
        Files.writeString(outPath, json.toString(), StandardCharsets.UTF_8); // Escribe el JSON completo en el archivo de salida en UTF-8

        Path csvPath = toCsvPath(outPath);                               // Genera la ruta del CSV reemplazando la extensión .json por .csv
        writeCsv(csvPath, rows);                                         // Escribe todos los resultados en formato CSV

        out.println("\n  Resultados guardados en:");                      // Informa que los archivos fueron guardados
        out.println("    JSON: " + outPath);                             // Muestra la ruta del JSON generado
        out.println("    CSV:  " + csvPath);                             // Muestra la ruta del CSV generado

        out.println("\n============================================================"); // Línea decorativa de cierre
        out.println("  BENCHMARK JAVA COMPLETADO");                                    // Mensaje final de éxito
        out.println("============================================================");   // Última línea decorativa
    }

    private static String esc(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\""); // Escapa barras invertidas y comillas dobles para producir JSON válido
    }

    private static String csvEsc(String s) {
        if (s == null) return "";                              // Si el valor es nulo, retorna cadena vacía para evitar NullPointerException
        boolean needsQuotes = s.contains(",") || s.contains("\"") || s.contains("\n") || s.contains("\r");
        // Determina si el valor necesita comillas: si contiene coma, comilla doble o saltos de línea
        String v = s.replace("\"", "\"\"");                   // Escapa las comillas dobles duplicándolas (estándar CSV RFC 4180)
        return needsQuotes ? ("\"" + v + "\"") : v;           // Envuelve en comillas si es necesario, o retorna el valor limpio
    }

    private static Path toCsvPath(Path jsonPath) {
        String name = jsonPath.getFileName().toString();       // Obtiene solo el nombre del archivo JSON sin la ruta completa
        String base = name.endsWith(".json") ? name.substring(0, name.length() - 5) : name;
        // Elimina la extensión .json del nombre si la tiene, para construir el nombre base sin extensión
        return jsonPath.resolveSibling(base + ".csv");         // Construye la ruta del CSV en el mismo directorio que el JSON
    }

    private static void writeCsv(Path path, List<ResultRow> rows) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("algorithm,language,n,timeSeconds,complexity\n"); // Escribe la fila de encabezados del CSV (sin columna timedOut porque búsqueda no tiene timeout)
        for (ResultRow r : rows) {                             // Itera sobre cada resultado para escribir una fila
            sb.append(csvEsc(r.algorithm())).append(",");      // Agrega el nombre del algoritmo escapado para CSV
            sb.append(csvEsc(r.language())).append(",");       // Agrega el lenguaje escapado para CSV
            sb.append(r.n()).append(",");                       // Agrega el tamaño del arreglo
            sb.append(String.format(Locale.US, "%.9f", r.timeSeconds())).append(","); // Agrega el tiempo con 9 decimales y punto decimal
            sb.append(csvEsc(r.complexity()));                 // Agrega la complejidad teórica escapada (última columna, sin coma al final)
            sb.append("\n");                                   // Salto de línea para la siguiente fila
        }
        Files.writeString(path, sb.toString(), StandardCharsets.UTF_8); // Escribe todo el contenido CSV en el archivo en UTF-8
    }

    private static int[] loadInts(Path path, int maxCount) throws IOException {
        int cap = maxCount == Integer.MAX_VALUE ? 1_000_000 : maxCount;
        // Si no hay límite (Integer.MAX_VALUE), usa 1.000.000 como capacidad inicial del buffer
        int[] buf = new int[Math.min(cap, 1_000_000)];        // Crea el buffer inicial con la capacidad calculada (máximo 1.000.000)
        int idx = 0;                                           // Índice para rastrear cuántos enteros se han leído
        try (BufferedReader br = Files.newBufferedReader(path, StandardCharsets.UTF_8)) { // Abre el archivo con buffer y UTF-8, se cierra automáticamente
            String line;
            while ((line = br.readLine()) != null) {           // Lee línea por línea hasta llegar al final del archivo
                line = line.trim();                            // Elimina espacios y saltos de línea alrededor del valor
                if (line.isEmpty()) continue;                  // Salta las líneas vacías sin procesarlas
                if (idx == buf.length) {                       // Si el buffer está lleno...
                    buf = Arrays.copyOf(buf, buf.length * 2); // ...lo duplica de tamaño para seguir leyendo (estrategia de crecimiento dinámico)
                }
                buf[idx++] = Integer.parseInt(line);           // Convierte la línea a entero y lo almacena en la siguiente posición del buffer
                if (idx >= maxCount) break;                    // Si se alcanzó el límite máximo de elementos, detiene la lectura
            }
        }
        return Arrays.copyOf(buf, idx);                        // Retorna una copia recortada al tamaño real de elementos leídos (sin posiciones vacías)
    }
}